packages/connectivity/src/scp-cf/resilience/resilience.ts

```diff
@@ -0,0 +1,233 @@
+import { createCircuitBreaker } from './circuit-breaker';
+import {
+  defaultResilienceOptions,
+  Middleware,
+  MiddlewareInOutOptions,
+  RequestHandler,
+  ResilienceMiddlewareOptions,
+  ResilienceOptions,
+  TimeoutOptions
+} from './resilience-options';
+
+interface ResilienceMiddlewareMapValue {
+  middleware: Middleware<any>;
+  options: Omit<ResilienceMiddlewareOptions, 'id'>;
+}
+const resilienceMiddlewareMap = new Map<string, ResilienceMiddlewareMapValue>();
+
+/**
+ * Create a promise for a time out race.
+ * @param timeout - Value for the timeout.
+ * @returns A promise which times out after the given time.
+ * @internal
+ */
+async function timeoutPromise<T>(timeout: number): Promise<T> {
+  return new Promise<T>((_, reject) =>
+    setTimeout(() => reject(new Error(`Timed out after ${timeout}ms`)), timeout)
+  );
+}
+
+/**
+ * Add time out to a function call.
+ * @param requestHandler - Function handler with empty parameter.
+ * @param timeout - t.
+ * @returns TODO: Add JSDoc later.
+ * @internal
+ */
+export function addTimeout<T>(
+  requestHandler: RequestHandler<T>,
+  timeout: TimeoutOptions
+): Promise<T> {
+  // If timeout is non-positive, we don't add timeout to the promise.
+  if (!timeout || timeout < 0) {
+    return requestHandler();
+  }
+  const racePromise = timeoutPromise<T>(timeout);
+  return Promise.race([requestHandler(), racePromise]);
+}
+
+/**
+ * TODO: Add JSDoc later.
+ * @param resilienceMiddlewareOptions - TODO: Add JSDoc later.
+ * @returns TODO: Add JSDoc later.
+ * @internal
+ */
+export function createTimeoutMiddleware<T>(
+  resilienceMiddlewareOptions: ResilienceMiddlewareOptions
+): Middleware<T> {
+  const { timeout } = resilienceMiddlewareOptions;
+  const circuitBreakerMiddlewareFn = (
+    middlewareInOutOptions: MiddlewareInOutOptions<T>
+  ) => {
+    const timeoutOptions = timeout();
+    return addTimeout(middlewareInOutOptions.fn, timeoutOptions);
+  };
+
+  return (middlewareInOutOptions: MiddlewareInOutOptions<T>) => ({
+    fn: () => circuitBreakerMiddlewareFn(middlewareInOutOptions),
+    exitChain: middlewareInOutOptions.exitChain
+  });
+}
+
+/**
+ * TODO: Add JSDoc later.
+ * @param resilienceMiddlewareOptions - TODO: Add JSDoc later.
+ * @returns TODO: Add JSDoc later.
+ * @internal
+ */
+export function createCircuitBreakerMiddleware<T>(
+  resilienceMiddlewareOptions: ResilienceMiddlewareOptions
+): Middleware<T> {
+  const { circuitBreaker } = resilienceMiddlewareOptions;
+  const timeoutMiddlewareFn = (
```

Comment:
[q] I guess this is a copy paste error and the name should be different.

Reply:
Good spot!